tool/src/java/mod.rs

```diff
@@ -0,0 +1,1667 @@
+pub(crate) struct MethodTpl<'a> {
+    method_name: Option<Cow<'a, str>>,
+    is_static: bool,
+    return_ty: Cow<'a, str>,
+    native_invocation: Cow<'a, str>,
+    params: Vec<Param<'a>>,
+    param_conversions: Vec<ParamConversion<'a>>,
+    return_conversion: Cow<'a, str>,
+    allocations: bool,
+    native_return_void: bool,
+}
+
+#[derive(Template, Clone, Debug)]
+#[template(path = "java/Struct.java.jinja", escape = "none")]
+pub(crate) struct StructTypeTpl<'a> {
+    type_name: Cow<'a, str>,
+    lib_name: Cow<'a, str>,
+    domain: Cow<'a, str>,
+    edges: Vec<Cow<'a, str>>,
+    fields: Vec<FieldTpl<'a>>,
+    methods: Vec<Cow<'a, str>>,
+    is_zst: bool,
+}
+
+#[derive(Clone, Debug)]
+struct FieldTpl<'a> {
+    name: Cow<'a, str>,
+    native_name: Cow<'a, str>,
+    to_native: Conversion<'a>,
+    to_java: Conversion<'a>,
+    ty: Cow<'a, str>,
+}
+
+#[derive(Template, Clone, Debug)]
+#[template(path = "java/Enum.java.jinja", escape = "none")]
+pub(crate) struct EnumTypeTpl<'a> {
+    type_name: Cow<'a, str>,
+    lib_name: Cow<'a, str>,
+    domain: Cow<'a, str>,
+    variants: Vec<VariantTpl<'a>>,
+    methods: Vec<Cow<'a, str>>,
+}
+
+#[derive(Clone, Debug)]
+struct VariantTpl<'a> {
+    name: Cow<'a, str>,
+    index: Cow<'a, str>,
+}
+
+#[derive(Template, Clone, Debug)]
+#[template(path = "java/Opaque.java.jinja", escape = "none")]
+pub(crate) struct OpaqueTypeTpl<'a> {
+    type_name: Cow<'a, str>,
+    lib_name: Cow<'a, str>,
+    domain: Cow<'a, str>,
+    edges: Vec<Cow<'a, str>>,
+    static_methods: Vec<Cow<'a, str>>,
+    class_methods: Vec<Cow<'a, str>>,
+}
+
+#[derive(Debug, Clone)]
+struct LifetimeTpl<'a> {
+    name: Cow<'a, str>,
+    edges: Vec<Cow<'a, str>>,
+}
+
+#[derive(Template, Clone, Debug)]
+#[template(path = "java/StructConversion.java.jinja", escape = "none")]
+pub(crate) struct StructConversionTpl<'a> {
+    lifetimes: Vec<LifetimeTpl<'a>>,
+    ty: Cow<'a, str>,
+    name: Cow<'a, str>,
+    native_val: Cow<'a, str>,
+}
+
+#[derive(Template, Clone, Debug)]
+#[template(path = "java/OpaqueConversion.java.jinja", escape = "none")]
+pub(crate) struct OpaqueConversionTpl<'a> {
+    name: Cow<'a, str>,
+    lifetimes: Vec<LifetimeTpl<'a>>,
+    self_edges: Option<Cow<'a, str>>,
+    ty: Cow<'a, str>,
+    native_val: Cow<'a, str>,
+}
+
+#[derive(Template, Clone, Debug)]
+#[template(path = "java/OpaqueConstructor.java.jinja", escape = "none")]
+pub(crate) struct OpaqueConstructorTpl<'a> {
+    return_ty: Cow<'a, str>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+struct Config<'cx> {
+    domain: Cow<'cx, str>,
+    lib_name: Cow<'cx, str>,
+}
+
+struct TyGenContext<'a, 'cx> {
+    tcx: &'cx TypeContext,
+    tcx_config: Config<'cx>,
+    formatter: &'a JavaFormatter<'cx>,
+    errors: &'a ErrorStore<'cx, String>,
+}
+
+#[derive(Debug, Clone)]
+struct Conversion<'cx> {
+    conversion: Cow<'cx, str>,
+    converted_value: Cow<'cx, str>,
+}
+
+mod arena {
```

Comment:
I wanted param conversions to not have to be aware of the context wherein they are done, and so introduced a kind of inversion of control, where if a conversion needs an arena it returns a variant that can only be accessed by providing an arena.